blog/en/blog/2022/08/17/backend-for-frontend-demo.md

```diff
@@ -0,0 +1,299 @@
+
+```yaml
+routes:
+  - uri: /                     #1
+    plugins:
+      bff: ~                   #2
+      public-api: ~            #2
+```
+
+1. For demo purposes, I preferred to set it at the root instead of `/bff/*`
+2. Declare the two plugins. Note that I'm using the [stand-alone mode](https://apisix.apache.org/docs/apisix/stand-alone/).
+
+First, we need to describe the plugin and not forget to return it at the end of the file:
+
+```lua
+local plugin_name = 'bff-plugin'
+
+local _M = {                           --1
+    version = 1.0,
+    priority = 100,                    --2
+    name = plugin_name,
+    schema = {},                       --3
+}
+
+return _M                              --4
+```
+
+1. The table needs to be named `_M`
+2. In this scenario, `priority` is irrelevant as no other plugins are involved (but `public_api`)
+3. No schema is necessary as there's no configuration
+4. Don't forget to return it!
+
+A plugin that has a public API needs to define an `api()` function returning an object describing matching HTTP methods, the matching URI, and the handler function.
+
+```lua
+function _M.api()
+    return {
+        {
+            methods = { 'GET' },
+            uri = "/",
+            handler = fetch_all_data,
+        }
+    }
+end
+```
+
+Now, we have to define the `fetch_all_data` function. It's only a matter of making HTTP calls to the catalog and newsfeed microservices. Have a look at [the code](https://github.com/ajavageek/backend-for-frontend/blob/master/bff_plugin/init.lua#L24-L52) if you're interested in the exact details.
+
+At this point, the (single) client can query `http://localhost:9080/` and get the complete payload.
+
+In a "real life" microservices-based organization, every team should be independent of each other. With this approach, each can develop its own BFF as a plugin and deploy it independently in the gateway.
+
+## Bonus: a poor man's BFF
+
+The microservices architecture creates two problems for clients:
+
+1. The need to fetch all data and filter out the unnecessary ones
+2. Multiple calls to each service
+
+The BFF pattern allows to fix both of them, at the cost of custom development, regardless whether it's for a dedicated app or a gateway plugin. If you're not willing to spend time in custom development, you can still avoid #2 by using a nifty plugin of Apache APISIX, `batch-requests`:
+
+>The `batch-requests` plugin accepts multiple requests, sends them from APISIX via HTTP pipelining, and returns an aggregated response to the client.
+>
+>This improves the performance significantly in cases where the client needs to access multiple APIs.
+>
+> -- [batch-requests](https://apisix.apache.org/docs/apisix/plugins/batch-requests/)
```

Comment:
```suggestion
:::note

The [`batch-requests`](https://apisix.apache.org/docs/apisix/plugins/batch-requests/) plugin accepts multiple requests, sends them from APISIX via HTTP pipelining, and returns an aggregated response to the client.

This improves the performance significantly in cases where the client needs to access multiple APIs.

:::
```

Reply:
You can use this format, the reference is as follows, and I incorporated the separate plugin link into the body.
<img width="847" alt="image" src="https://user-images.githubusercontent.com/39793568/185271125-e204d7a3-1d4b-43fa-8f4c-656419a60a96.png">

Reply:
Actually, it's not a note but a direct copy-paste from the documentation. Hence, it makes sense to use a quote.

Reply:
Got it, then just ignore this comment^^